sql/schemachanger: Fix early exit for IF NOT EXISTS when adding existing columns

spilchen opened the pull request:
Previously, we exited too early when adding an existing column with the IF NOT EXISTS option. This skipped necessary AST handling, such as annotating resolved names, which subsequent processing depended on. To fix this, the existing column check has been moved to just before adding or removing SCPB elements.

Additionally, we failed to handle cases where the column had been dropped in a prior statement of an explicit transaction. Previously, the operation would succeed without adding any elements. Now, the column will be correctly added.

Epic: None
Closes #137326
Release note (bug fix): Fixed an issue where adding an existing column with the IF NOT EXISTS option could exit too early, skipping necessary handling of the AST. This could lead to statement failure of the ALTER.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137633)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice work tracking this down. let's backport this one if it's relatively easy.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137633#-:-OELYzSD2JVeT58kP9lF:bjum8rq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Fix early exit for IF NOT EXISTS when adding existing columns

Previously, we exited too early when adding an existing column with the
IF NOT EXISTS option. This skipped necessary AST handling, such as
annotating resolved names, which subsequent processing depended on. To
fix this, the existing column check has been moved to just before adding
or removing SCPB elements.

Additionally, we failed to handle cases where the column had been
dropped in a prior statement of an explicit transaction. Previously, the
operation would succeed without adding any elements. Now, the column will be
correctly added.

Epic: None
Closes #137326
Release note (bug fix): Fixed an issue where adding an existing column
with the IF NOT EXISTS option could exit too early, skipping necessary
handling of the AST. This could lead to statement failure of the ALTER.

spilchen commented:
TFTR!

bors r+
